### ISSUE TITLE
TS ProgramGen introduces unnecessary applies for array indexing

### ISSUE THREAD

**Issue by iwahbe** [pulumi/pulumi#10543]
### What happened?

Running `pulumi convert` on a Pulumi YAML produced unnecessarily messy output.

### Steps to reproduce

Run `pulumi convert` on 
```yaml
name: test-acm
runtime: yaml
description: A minimal AWS Pulumi YAML program

variables:
  zoneId:
    Fn::Invoke:
      Function: aws:route53:getZone
      Arguments:
        name: "example.com"
        privateZone: false
      Return: id
resources:
  exampleCertificate:
    type: aws:acm:Certificate
    properties:
      domainName: "example.com"
      validationMethod: "DNS"
  certValidation:
    type: aws:route53:Record
    properties:
      name: ${exampleCertificate.domainValidationOptions[0].resourceRecordName}
      records:
        - ${exampleCertificate.domainValidationOptions[0].resourceRecordValue}
      ttl: 60
      type: ${exampleCertificate.domainValidationOptions[0].resourceRecordType}
      zoneId: ${zoneId}
  certCertificateValidation:
    type: aws:acm:CertificateValidation
    properties:
      certificateArn: ${exampleCertificate.arn}
      validationRecordFqdns:
        - ${certValidation.fqdn}
outputs:
  certificateArn: ${certCertificateValidation.certificateArn}
```

### Expected Behavior

We should generate the following (or something equivalent):
```typescript
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const zoneId = aws.route53.getZone({
    name: "example.com",
    privateZone: false,
}).then(invoke => invoke.id);
const exampleCertificate = new aws.acm.Certificate("exampleCertificate", {
    domainName: "example.com",
    validationMethod: "DNS",
});
const certValidation = new aws.route53.Record("certValidation", {
    name: exampleCertificate.domainValidationOptions[0].resourceRecordName,
    records: [exampleCertificate.domainValidationOptions[0].resourceRecordValue],
    ttl: 60,
    type: exampleCertificate.domainValidationOptions[0].resourceRecordType,
    zoneId: zoneId,
});
const certCertificateValidation = new aws.acm.CertificateValidation("certCertificateValidation", {
    certificateArn: exampleCertificate.arn,
    validationRecordFqdns: [certValidation.fqdn],
});
export const certificateArn = certCertificateValidation.certificateArn;
```

### Actual Behavior

We generate
```typescript
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const zoneId = aws.route53.getZone({
    name: "example.com",
    privateZone: false,
}).then(invoke => invoke.id);
const exampleCertificate = new aws.acm.Certificate("exampleCertificate", {
    domainName: "example.com",
    validationMethod: "DNS",
});
const certValidation = new aws.route53.Record("certValidation", {
    name: exampleCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordName),
    records: [exampleCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
    ttl: 60,
    type: exampleCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordType).apply((x) => aws.route53.recordtype.RecordType[x]),
    zoneId: zoneId,
});
const certCertificateValidation = new aws.acm.CertificateValidation("certCertificateValidation", {
    certificateArn: exampleCertificate.arn,
    validationRecordFqdns: [certValidation.fqdn],
});
export const certificateArn = certCertificateValidation.certificateArn;
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by iwahbe** [comment]
Do we want to consider this as part of the `pulumi convert` work before YAML GA? Right now its not a great result, but is technically correct, so I'm unsure.
CC @AaronFriel @aq17

**Comment by aq17** [comment]
> Do we want to consider this as part of the `pulumi convert` work before YAML GA? Right now its not a great result, but is technically correct, so I'm unsure. CC @AaronFriel @aq17

@iwahbe  This seems like it would require us to introduce variables not in the original YAML, such as `exampleZone`, and omit others such as `zoneId`? If we limit this to when we see that a variable is used only once as an input to another variable/ resource, it might be worth it for readability?

**Comment by iwahbe** [comment]
My bad @aq17. I miscopied. The problem I wanted to address is that we generate lots of extra `apply`s when we don't need them. I have updated the "Expected Behavior" and "Actual Behavior" sections appropriately.

**Comment by aq17** [comment]
@iwahbe Ah I see. I think it'd be a nice improvement but will defer to @AaronFriel to determine if it's worth prioritizing right now

**Comment by AaronFriel** [comment]
I think it's worth checking to see how large it is, but it's lower priority than any of the `convert` issues that block programs. 

We already have a programgen pass that rewrites applies on outputs for property accesses. Based on this comment, I'm wondering if there was a regression and we previously supported this:

https://github.com/pulumi/pulumi/blob/047c429223074bcc1110339079fcbe813216cca6/pkg/codegen/nodejs/gen_program_lower.go#L65-L71

**Comment by mikhailshilkov** [comment]
Programgen quality (unrelated to YAML) is a long-due investment area that we might finally take more seriously in Q4. I'd say we can park this ticket until it's ranked against the other programgen issues.
